internal/app/machined/internal/server/v1alpha1/v1alpha1_server.go

```diff
@@ -505,6 +505,147 @@ func (s *Server) List(req *machine.ListRequest, obj machine.MachineService_ListS
 	return nil
 }
 
+// DiskUsage implements the machine.MachineServer interface.
+func (s *Server) DiskUsage(req *machine.DiskUsageRequest, obj machine.MachineService_DiskUsageServer) error { //nolint: gocyclo
+	if req == nil {
+		req = new(machine.DiskUsageRequest)
+	}
+
+	for _, path := range req.Paths {
+		if !strings.HasPrefix(path, OSPathSeparator) {
+			// Make sure we use complete paths
+			path = OSPathSeparator + path
+		}
+
+		path = strings.TrimSuffix(path, OSPathSeparator)
+		if path == "" {
+			path = "/"
+		}
+
+		_, err := os.Stat(path)
+		if err == os.ErrNotExist {
+			return fmt.Errorf("%s no such file or directory", path)
+		}
+
+		files, err := archiver.Walker(obj.Context(), path, archiver.WithMaxRecurseDepth(-1))
+		if err != nil {
+			return err
+		}
+
+		folders := map[string]*machine.DiskUsageInfo{}
+
+		// send a record back to client if the message shouldn't be skipped
+		// at the same time use record information for folder size estimation
+		sendSize := func(info *machine.DiskUsageInfo, depth int32, isDir bool) {
+			prefix := strings.TrimRight(filepath.Dir(info.Name), "/")
+			if folder, ok := folders[prefix]; ok {
+				folder.Size += info.Size
+			}
+
+			// recursion depth check
+			skip := depth >= req.RecursionDepth && req.RecursionDepth > 0
+			// skip files check
+			skip = skip || !isDir && !req.All
+			// threshold check
+			skip = skip || req.Threshold > 0 && info.Size < req.Threshold
+			skip = skip || req.Threshold < 0 && info.Size > -req.Threshold
+
+			if skip {
+				return
+			}
+
+			err = obj.Send(info)
```

Comment:
once we send the info back to the client, should we remove it from the `folders` map to control the memory usage?